sql/core/src/test/scala/org/apache/spark/sql/execution/RemoveRedundantProjectsSuite.scala

```diff
@@ -215,6 +217,27 @@ abstract class RemoveRedundantProjectsSuiteBase
         |LIMIT 10
         |""".stripMargin
     assertProjectExec(query, 0, 3)
+
+  }
+  Seq("true", "false").foreach { codegenEnabled =>
+    test("SPARK-35287: project generating unsafe row " +
+      s"should not be removed (codegen=$codegenEnabled)") {
+      withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1",
```

Comment:
Why do we need to set the broadcast hash join threshold and the leaf node default parallelism?

Reply:
Please see the comment added.

Reply:
If the pattern SortExec -> BatchScamExec can trigger this issue, would an orderBy instead of a join be sufficient?

Reply:
orderBy can trigger this issue and choose it for test.